8344140: Refactor the discovery of AOT cache artifacts

iklam opened the pull request:
This is a clean up after the initial integration for [JEP 483](https://bugs.openjdk.org/browse/JDK-8315737)

- Merged 3 loops into 1 for discovering the classes and oops that should be stored in the AOT cache. About 250 lines of code are deleted.
- Added comments in aotArtifactFinder.hpp to describe the logic, which is much simpler than before.
- Enum classes are no longer unconditionally AOT-initialized -- an Enum class is AOT-initialized only if we find an archived oop of this type.

Note to reviewers:

One consequence of this refactoring is that archived oops are now discovered (by heapShared.cpp) before the metadata are copied (by ArchiveBuilder). There are some functions that depend on the old order (metadata were copied before oop discovery), so I had to shuffle them around. Examples are `Modules::check_archived_module_oop()`, or the new code in `Klass::remove_java_mirror()`.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8344140](https://bugs.openjdk.org/browse/JDK-8344140): Refactor the discovery of AOT cache artifacts (**Enhancement** - P4)


### Reviewers
 * [Ashutosh Mehra](https://openjdk.org/census#asmehra) (@ashu-mehra - Committer) Review applies to [c5ed9578](https://git.openjdk.org/jdk/pull/22291/files/c5ed9578642ae120ec96b7d838de7afcbf2a3616)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22291/head:pull/22291` \
`$ git checkout pull/22291`

Update a local copy of the PR: \
`$ git checkout pull/22291` \
`$ git pull https://git.openjdk.org/jdk.git pull/22291/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22291`

View PR using the GUI difftool: \
`$ git pr show -t 22291`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22291.diff">https://git.openjdk.org/jdk/pull/22291.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22291#issuecomment-2490117257)
</details>

ashu-mehra commented:
@iklam nice cleanup! The comments and the logic in AOTArtifactFinder makes it a lot easier to understand the algorithm now.

ashu-mehra commented:
Comment in `AOTClassInitializer::can_archive_initialized_mirror()` refers to `HeapShared::find_all_aot_initialized_classes()` which does not exist anymore:

```
  // When CDSConfig::is_initing_classes_at_dump_time() is enabled,
  // HeapShared::find_all_aot_initialized_classes() finds the classes of all
  // heap objects that are reachable from HeapShared::_run_time_special_subgraph,
  // and mark these classes as aot-inited.
```

And this comment at line 263 in `aotClassInitializer.cpp` also needs to be updated as the enum classes are not unconditionally AOT-initialized:

```
      // above we selected all enums; we must include their super as well
      {"java/lang/Enum"},
```

iklam commented:
> Comment in `AOTClassInitializer::can_archive_initialized_mirror()` refers to `HeapShared::find_all_aot_initialized_classes()` which does not exist anymore:
> 
> ```
>   // When CDSConfig::is_initing_classes_at_dump_time() is enabled,
>   // HeapShared::find_all_aot_initialized_classes() finds the classes of all
>   // heap objects that are reachable from HeapShared::_run_time_special_subgraph,
>   // and mark these classes as aot-inited.
> ```
> 
> And this comment at line 263 in `aotClassInitializer.cpp` also needs to be updated as the enum classes are not unconditionally AOT-initialized:
> 
> ```
>       // above we selected all enums; we must include their super as well
>       {"java/lang/Enum"},
> ```

I fixed the comments and removed the entry for `java/lang/Enum`

iklam commented on src/hotspot/share/cds/metaspaceShared.cpp:
```diff
@@ -543,6 +544,7 @@ class StaticArchiveBuilder : public ArchiveBuilder {
   virtual void iterate_roots(MetaspaceClosure* it) {
     FileMapInfo::metaspace_pointers_do(it);
+    AOTArtifactFinder::all_cached_classes_do(it);
     SystemDictionaryShared::dumptime_classes_do(it);
```
The call to `SystemDictionaryShared::dumptime_classes_do()` is still necessary. It does more than just discovering the classes. It also walks the `DumpTimeClassInfo` with

```
info.metaspace_pointers_do(it);
```

which will walk the loader constraints, etc.

ashu-mehra replied:
Right, missed that. Thanks for pointing that out.

calvinccheung commented on src/hotspot/share/cds/aotArtifactFinder.hpp:
```diff
@@ -0,0 +1,90 @@
+//
+// Note1: See TODO comments in HeapShared::archive_object() for exceptions to this rule.
+//
+// Note2: The scanning of Java objects is done in heapShared.cpp. Please see calls into the HeapShared class
```
Note2 isn't referenced by the comment above. Perhaps line 62 should reference it?

calvinccheung commented on src/hotspot/share/cds/heapShared.cpp:
```diff
@@ -311,6 +314,36 @@ bool HeapShared::archive_object(oop obj) {
+    Klass* k = obj->klass();
+    if (k->is_instance_klass()) {
+      // Whenever we see a non-array Java object of type X, we mark X to be aot-initialized.
+      // This ensures that during the production run, whenever Java code seens a cached object
```
typo: s/seens/sees

iklam replied:
Fixed.

calvinccheung commented on src/hotspot/share/cds/heapShared.cpp:
```diff
@@ -311,6 +314,36 @@ bool HeapShared::archive_object(oop obj) {
+        // We can't rerun <clinit> of enum classes (see cdsEnumKlass.cpp) so
+        // we must store them as AOT-initialized.
+        AOTArtifactFinder::add_aot_inited_class(InstanceKlass::cast(k));
+      } else if (subgraph_info == _dump_time_special_subgraph) {
```
Can this be an "or" condition rather than "else if" since the same function is called for either on of the conditions?

iklam replied:
Fixed.

calvinccheung commented on src/hotspot/share/cds/heapShared.cpp:
```diff
@@ -1082,29 +768,26 @@ void KlassSubGraphInfo::add_subgraph_entry_field(int static_field_offset, oop v)
 
-  if (buffered_k->is_instance_klass()) {
+  if (orig_k->is_instance_klass()) {
+    InstanceKlass* ik = InstanceKlass::cast(orig_k);
```
Should this be inside a "#if ASSERT" since `ik` is used only in the `assert` statements?

iklam replied:
Fixed.

calvinccheung reviewed (commented):
Some minor comments below.

calvinccheung reviewed (approved):
Updates look good.

iklam commented:
Thanks @ashu-mehra and @calvinccheung for the review.
I've tested against tiers 1~4 and builds-tier5. No new failures.
/integrate
